# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [263]:
# importando biblioteca pandas
import pandas as pd
# leitura da base demo01.csv
demo01 = pd.read_csv('demo01.csv')

In [265]:
demo01.head(n = 5)

sexo posse_de_veiculo posse_de_imovel  qtd_filhos            tipo_renda  \
0    M                Y               Y           0               Working   
1    F                N               Y           0  Commercial associate   
2    F                N               Y           0  Commercial associate   
3    M                Y               Y           0               Working   
4    F                Y               N           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   

       idade  tempo_emprego  possui_celular  possui_fone_comercial  \
0  58.832877       3.106849               1                      0   
1  52.356164       8.358904               1                      0   
2  52.356164       8.358904               1                      0   
3  46.224658       2.106849               1                      1   
4  29.230137       3.021918               1                      0   

   possui_fone  possui_email  qt_pessoas_residencia    mau  
0            0             0                    2.0  False  
1            1             1                    1.0  False  
2            1             1                    1.0  False  
3            1             1                    2.0  False  
4            0             0                    2.0  False

In [267]:
# mapeamento das principais informações do DataFrame pelo método .info()
demo01.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   sexo                   16650 non-null  object 
 1   posse_de_veiculo       16650 non-null  object 
 2   posse_de_imovel        16650 non-null  object 
 3   qtd_filhos             16650 non-null  int64  
 4   tipo_renda             16650 non-null  object 
 5   educacao               16650 non-null  object 
 6   estado_civil           16650 non-null  object 
 7   tipo_residencia        16650 non-null  object 
 8   idade                  16650 non-null  float64
 9   tempo_emprego          16650 non-null  float64
 10  possui_celular         16650 non-null  int64  
 11  possui_fone_comercial  16650 non-null  int64  
 12  possui_fone            16650 non-null  int64  
 13  possui_email           16650 non-null  int64  
 14  qt_pessoas_residencia  16650 non-null  float64
 15  ma

In [269]:
# mapeamento dos tipos das variáveis pelo método .dtypes
demo01.dtypes

sexo                      object
posse_de_veiculo          object
posse_de_imovel           object
qtd_filhos                 int64
tipo_renda                object
educacao                  object
estado_civil              object
tipo_residencia           object
idade                    float64
tempo_emprego            float64
possui_celular             int64
possui_fone_comercial      int64
possui_fone                int64
possui_email               int64
qt_pessoas_residencia    float64
mau                         bool
dtype: object

In [271]:
# quantidade de linhas e colunas pelo método .shape
demo01.shape

(16650, 16)

In [273]:
# quantidade de valores missing pelo método .isna()
demo01.isna().sum()

sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
possui_celular           0
possui_fone_comercial    0
possui_fone              0
possui_email             0
qt_pessoas_residencia    0
mau                      0
dtype: int64

In [275]:
# remover dados duplicados pelo método .drop_duplicates()
demo01_sem_dup = demo01.drop_duplicates()

In [277]:
# quantidade de linhas e colunas pelo método .shape após o uso do método .drop_duplicates()
print("Quantidade de linhas e colunas sem o método .drop_duplicates():", demo01.shape)
print("Quantidade de linhas e colunas sem o método .drop_duplicates():", demo01_sem_dup.shape)

Quantidade de linhas e colunas sem o método .drop_duplicates(): (16650, 16)
Quantidade de linhas e colunas sem o método .drop_duplicates(): (6770, 16)


In [279]:
# resetar indices pelo método .reset_index()
demo01_sem_dup.reset_index(drop = True, inplace = True)

In [281]:
# distribuição dos valores da variável 'mau' pelo método .value_counts()
demo01_sem_dup['mau'].value_counts()

mau
False    6472
True      298
Name: count, dtype: int64

#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável.
    Dica:
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis,
        3. conte o número de categorias dessa variável
        4. acumule essa informação de 3. na lista que você criou em 1.
        5. No final, essa lista pode ser atribuída à nossa variável.

In [283]:
# criar DataFrame vazio denominado 'metadados'
metadados = pd.DataFrame()

In [285]:
# dict com o nome e o tipo das variáveis do DataFrame 'demo01'
dict = {'nome_variavel':['sexo', 'posse_de_veiculo', 'posse_de_imovel', 'qtd_filhos', 'tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia',
                        'idade', 'tempo_emprego', 'possui_celular', 'possui_fone_comercial', 'possui_fone', 'possui_email',
                        'qt_pessoas_residencia', 'mau'],
       'tipo_variavel':['object', 'object', 'object', 'int64', 'object', 'object', 'object', 'object', 'float64', 'float64', 'int64', 'int64',
                       'int64', 'int64', 'float64', 'bool']}

In [287]:
# adicionar o dict 'nomes_variaveis' dentro do DataFrame 'metadados'
metadados = pd.DataFrame(data = dict)

In [291]:
# declaração das variáveis, Series e lista que serão utilizadas no loop while
i = 0
j = (len(metadados['nome_variavel']) - 1)
series_nome_variavel = metadados['nome_variavel']
lista_valores_variaveis = []

# loop while começando em i e terminando em j
while i <= j:
    posicao = series_nome_variavel[i] # retorna o nome da variável
    valor = demo01_sem_dup[posicao].nunique() # retorna o valor unico da variável mapeada acima
    lista_valores_variaveis.append(valor) # adiciona o valor na lista 'lista_valores_variaveis'
    i += 1

In [295]:
# lista 'lista_valores_variaveis' como coluna 'qtd_categorias' no DataFrame 'metadados'
metadados['qtd_categorias'] = lista_valores_variaveis

In [297]:
metadados

nome_variavel tipo_variavel  qtd_categorias
0                    sexo        object               2
1        posse_de_veiculo        object               2
2         posse_de_imovel        object               2
3              qtd_filhos         int64               8
4              tipo_renda        object               5
5                educacao        object               5
6            estado_civil        object               5
7         tipo_residencia        object               6
8                   idade       float64            5298
9           tempo_emprego       float64            3005
10         possui_celular         int64               1
11  possui_fone_comercial         int64               2
12            possui_fone         int64               2
13           possui_email         int64               2
14  qt_pessoas_residencia       float64               9
15                    mau          bool               2

#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [299]:
print('valores únicos da variável sexo: ', demo01_sem_dup['sexo'].unique())
print('valores únicos da variável posse de veiculo: ', demo01_sem_dup['posse_de_veiculo'].unique())
print('valores únicos da variável posse de imovel: ', demo01_sem_dup['posse_de_imovel'].unique())
print('valores únicos da variável tipo de renda: ', demo01_sem_dup['tipo_renda'].unique())
print('valores únicos da variável educação: ', demo01_sem_dup['educacao'].unique())
print('valores únicos da variável estado civil: ', demo01_sem_dup['estado_civil'].unique())
print('valores únicos da variável tipo de residência: ', demo01_sem_dup['tipo_residencia'].unique())

valores únicos da variável sexo:  ['M' 'F']
valores únicos da variável posse de veiculo:  ['Y' 'N']
valores únicos da variável posse de imovel:  ['Y' 'N']
valores únicos da variável tipo de renda:  ['Working' 'Commercial associate' 'State servant' 'Pensioner' 'Student']
valores únicos da variável educação:  ['Secondary / secondary special' 'Higher education' 'Incomplete higher'
 'Lower secondary' 'Academic degree']
valores únicos da variável estado civil:  ['Married' 'Single / not married' 'Civil marriage' 'Widow' 'Separated']
valores únicos da variável tipo de residência:  ['House / apartment' 'Rented apartment' 'Municipal apartment'
 'With parents' 'Co-op apartment' 'Office apartment']


In [301]:
# método map para conversão das variáveis 'sexo', 'posse_de_veiculos' e 'posse_de_imovel' em 0 e 1 
demo01_sem_dup.sexo = demo01_sem_dup.sexo.map({'M':1, 'F':0})
demo01_sem_dup.posse_de_veiculo = demo01_sem_dup.posse_de_veiculo.map({'Y':1, 'N':0})
demo01_sem_dup.posse_de_imovel = demo01_sem_dup.posse_de_imovel.map({'Y':1, 'N':0})

C:\Users\Leonardo Bariani\AppData\Local\Temp\ipykernel_25268\2094792353.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demo01_sem_dup.sexo = demo01_sem_dup.sexo.map({'M':1, 'F':0})
C:\Users\Leonardo Bariani\AppData\Local\Temp\ipykernel_25268\2094792353.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  demo01_sem_dup.posse_de_veiculo = demo01_sem_dup.posse_de_veiculo.map({'Y':1, 'N':0})
C:\Users\Leonardo Bariani\AppData\Local\Temp\ipykernel_25268\2094792353.py:4: SettingWithCopyWarning: 
A value is tr

In [305]:
# método pd.get_dummies() para criação de novas colunas com valores booleanos 'True' ou 'False' das variáveis 'tipo_renda', 'educacao', 'estado_civil' e 'tipo_residencia'
demo01_sem_dup = pd.get_dummies(demo01_sem_dup, columns = ['tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia'])

In [317]:
demo01_sem_dup.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6770 entries, 0 to 6769
Data columns (total 33 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   sexo                                    6770 non-null   int64  
 1   posse_de_veiculo                        6770 non-null   int64  
 2   posse_de_imovel                         6770 non-null   int64  
 3   qtd_filhos                              6770 non-null   int64  
 4   idade                                   6770 non-null   float64
 5   tempo_emprego                           6770 non-null   float64
 6   possui_celular                          6770 non-null   int64  
 7   possui_fone_comercial                   6770 non-null   int64  
 8   possui_fone                             6770 non-null   int64  
 9   possui_email                            6770 non-null   int64  
 10  qt_pessoas_residencia                   6770 non-null   floa

#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [334]:
# filtro mau = True = 1 no DataFrame 'demo01_sem_dup'
demo01_sem_dup_mau_pagador = demo01_sem_dup.query('mau == True')

# tabela cruzada das variáveis 'posse_de_veiculo' e 'possui_email'
tabela_mau_pagador = pd.crosstab(demo01_sem_dup_mau_pagador['posse_de_veiculo'], demo01_sem_dup_mau_pagador['possui_email'])
tabela_mau_pagador

possui_email        0   1
posse_de_veiculo         
0                 182   8
1                  96  12

A variável 'possui_email' parece mais poderosa para explicar os mau pagadores. Dos 298 registros de mau pagadores, 278 registros não possuem e-mail - 93% do total da base.

#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [339]:
demo01_sem_dup.to_csv('demo01_sem_dup.csv')